aws/s3/s3_concurrent.go

```diff
@@ -165,33 +190,58 @@ func (s *S3Concurrent) GetAllConcurrently(bucket, version string, objects []type
 	return s.manager.Process(processFunc, objects)
 }
 
-// getWorker retrieves a worker from the manager's worker pool.
-func (cm *ConcurrencyManager) getWorker() *worker {
-	return <-cm.workerPool
+// getWorker retrieves a number of workers from the manager's worker pool.
+func (cm *ConcurrencyManager) getWorkers(number int) []*worker {
+	cm.workerPool.mutex.Lock()
+	defer cm.workerPool.mutex.Unlock()
+
+	workers := make([]*worker, number)
+	for i := 0; i < number; i++ {
+		workers[i] = <-cm.workerPool.channel
+	}
+	return workers
 }
 
 // returnWorker returns a worker to the manager's worker pool.
 func (cm *ConcurrencyManager) returnWorker(w *worker) {
-	cm.workerPool <- w
+	cm.workerPool.channel <- w
 }
 
-// secureMemory secures enough memory from the manager's memory pool
-// to fit the specified size.
-func (cm *ConcurrencyManager) secureMemory(size int64) {
+// secureMemory secures the memory needed for the given objects
+// from the manager's memory pool.
+func (cm *ConcurrencyManager) secureMemory(objects []types.Object) {
+	cm.memoryPool.mutex.Lock()
+	defer cm.memoryPool.mutex.Unlock()
 
-	var securedMemory int64 = 0
-	for securedMemory < size {
-		securedMemory += <-cm.memoryPool
+	for _, o := range objects {
+		var securedMemory int64 = 0
+		for securedMemory < o.Size {
+			securedMemory += <-cm.memoryPool.channel
+		}
 	}
 }
 
+// calculateRequiredMemoryFor calculates the amount of memory required to contain
+// the given objects based on size. Useful as a precheck before securing to
+// ensure there's enough in the pool to fulfill the request.
+func (cm *ConcurrencyManager) calculateRequiredMemoryFor(objects []types.Object) int64 {
+	var totalMemory int64 = 0
+	for _, o := range objects {
+		var memoryForObject int64 = 0
+		for memoryForObject < o.Size {
```

Comment:
Since memoryChungSize is fixed, is there an easier math so we don't need a loop? 
```
chunks := math.Ceil(float64(o.Size) / float64(cm.memoryChungSize))
totalMemoy += cm.memoryChunkSize * chunks
```

Reply:
woops I didn't see your code suggestion. Mine is a bit longer but no floats.